tests/reconfiguration.py

```diff
@@ -610,6 +611,143 @@ def run_join_old_snapshot(args):
+                assert n.node_id in config_before["nodes"]
+
+        nodes_before = get_current_nodes_table(network)
+        for n in network.nodes:
+            assert n.node_id in nodes_before
+            assert nodes_before[n.node_id]["status"] == "Trusted"
+
+        # Check that the service config agrees that this is a 1tx network
+        with primary.client() as c:
+            s = c.get("/node/service/configuration").body.json()
+            assert s["reconfiguration_type"] == "OneTransaction"
+
+        network.consortium.submit_2tx_migration_proposal(primary)
+        network.wait_for_all_nodes_to_commit(primary)
+
+        # Check that the service config has been updated
+        with primary.client() as c:
+            rj = c.get("/node/service/configuration").body.json()
+            assert rj["reconfiguration_type"] == "TwoTransaction"
+
+        # Check that all nodes have updated their consensus parameters
+        for node in network.nodes:
+            with node.client() as c:
+                rj = c.get("/node/consensus").body.json()
+                assert "reconfiguration_type" in rj["details"]
+                assert rj["details"]["reconfiguration_type"] == "TwoTransaction"
+                assert len(rj["details"]["learners"]) == 0
+
+        # Add a new node and check that the reconfiguration is recorded correctly
+        tx_id_before = TxID(0, 0)
+        with primary.client() as c:
+            rj = c.get("/node/commit").body.json()
+            tx_id_before = TxID.from_str(rj["transaction_id"])
+
+        new_node = network.create_node("local://localhost")
+        network.join_node(new_node, args.package, args, from_snapshot=False)
+        network.trust_node(new_node, args)
+        network.wait_for_all_nodes_to_commit(primary)
+
+        config_after = get_current_config(network)
+        nodes_after = get_current_nodes_table(network)
+
+        assert len(nodes_after) == len(nodes_before) + 1
+        assert nodes_after[new_node.node_id]["status"] == "Trusted"
+
+        assert config_before is None or (
+            config_after["rid"] == config_before["rid"] + 1
+        )
+        assert new_node.node_id in config_after["nodes"]
+
+        # Check that the new node has the right consensus parameter
+        with new_node.client() as c:
+            rj = c.get("/node/consensus").body.json()
+            assert "reconfiguration_type" in rj["details"]
+            assert "learners" in rj["details"]
+            assert rj["details"]["reconfiguration_type"] == "TwoTransaction"
+            assert len(rj["details"]["learners"]) == 0
+
+        tx_id_after = TxID(0, 0)
+        with primary.client() as c:
+            rj = c.get("/node/commit").body.json()
+            tx_id_after = TxID.from_str(rj["transaction_id"])
+        pending_at, learner_at, trusted_at = get_joining_node_dates(
+            network, tx_id_before, tx_id_after, new_node.node_id
+        )
+        assert pending_at < learner_at < trusted_at
```

Comment:
I feel like this tests more than just the migration but also the generic learner addition process. Is this done here because it is hard to observe that a new node was added as a learner (because nodes catch up fast!)?

Reply:
Yes, this tests that the learner was in fact first added as a learner and then promoted to trusted/active, and that this happened between the two reconfiguration transactions. Since that's the basic 2tx guarantee, I wanted to make sure we test this. Like I said in the other comment above, maybe I should read the actual ledger files instead to make sure this is correctly auditable.